sdk/core/core/inc/az_json.h

```diff
@@ -24,28 +24,42 @@
 
 #include <_az_cfg_prefix.h>
 
-/*
- * @brief az_json_token_kind is an enum defining symbols for the various kinds of JSON tokens.
+/**
+ * @brief Defines symbols for the various kinds of JSON tokens that make up any JSON text.
  */
 typedef enum
 {
-  AZ_JSON_TOKEN_NULL,
-  AZ_JSON_TOKEN_BOOLEAN,
-  AZ_JSON_TOKEN_NUMBER,
-  AZ_JSON_TOKEN_STRING,
-  AZ_JSON_TOKEN_OBJECT,
-  AZ_JSON_TOKEN_OBJECT_START,
-  AZ_JSON_TOKEN_OBJECT_END,
-  AZ_JSON_TOKEN_ARRAY_START,
-  AZ_JSON_TOKEN_ARRAY_END,
-  // AZ_JSON_TOKEN_SPAN represents a token consisting of nested JSON; the JSON parser never returns
-  // a token of this type.
-  AZ_JSON_TOKEN_SPAN,
+  AZ_JSON_TOKEN_NONE, ///< There is no value (as distinct from #AZ_JSON_TOKEN_NULL).
+  AZ_JSON_TOKEN_OBJECT_START, ///< The token kind is the start of a JSON object.
+  AZ_JSON_TOKEN_OBJECT_END, ///< The token kind is the end of a JSON object.
+  AZ_JSON_TOKEN_ARRAY_START, ///< The token kind is the start of a JSON array.
+  AZ_JSON_TOKEN_ARRAY_END, ///< The token kind is the end of a JSON array.
+  AZ_JSON_TOKEN_PROPERTY_NAME, ///< The token kind is a JSON property name.
+  AZ_JSON_TOKEN_STRING, ///< The token kind is a JSON string.
+  AZ_JSON_TOKEN_NUMBER, ///< The token kind is a JSON number.
+  AZ_JSON_TOKEN_TRUE, ///< The token kind is the JSON literal `true`.
```

Comment:
Can you elaborate more on why we consider true/false sperate types.  Most developers think of true/false as a boolean

Reply:
> AZ_JSON_TOKEN_OBJECT is used to append one entire object to json.
We want to support customer creating one json and then adding it to other jsons. That's where we use that token for.

See PR description. 

> Post-preview: Decide whether we want to add support for writing an existing JSON within an `az_span`, raw, without any validation.

Deferring that feature to outside this PR as it require design consideration and I would prefer to avoid it if possible.

> Can you elaborate more on why we consider true/false sperate types. Most developers think of true/false as a boolean

Whether we have `boolean` or separate enum values, one component or the other has to add extra if-checks to differentiate between the two. Separating the two keeps the token distinct and now the parser (which already know that information), can relay that back to the user. Hiding that info within a `boolean` token kind means the caller has to check again which of the two bool values that token represents (for example, is it's length 4 or 5, etc.). I believe that is a more common operation, so I tried to avoid having the customer add a check in that scenario over others.

Here's the parser already needing to know the exact token type anyway:
https://github.com/Azure/azure-sdk-for-c/blob/d691e0fbcfcc1ff8de3b86632f02e1884fbae329/sdk/core/core/src/az_json_parser.c#L360-L367

Here's an example code snippet that illustrates the point a bit.

Before:
https://github.com/Azure/azure-sdk-for-c/blob/b88429a5293ea1c6f909cad2d1ee063e734acc39/sdk/core/core/test/cmocka/test_az_json.c#L781-L789

After:
https://github.com/Azure/azure-sdk-for-c/blob/d691e0fbcfcc1ff8de3b86632f02e1884fbae329/sdk/core/core/test/cmocka/test_az_json.c#L666-L681

There is no right answer here and is primarily a design choice based on perspective/judgement call. The JSON spec and domain doesn't map exactly to various programming language concepts. For example, the JSON number isn't what we would consider a `double` or an `int32_t` in C. So, the current design tries to capture the JSON spec and hence uses the terminology of that domain, which thinks of booleans as the `true` and `false` literals (similar to the `null` literal). The way I perceive this enum is that it answers the question: `What types of tokens does JSON contain?`

https://tools.ietf.org/html/rfc8259#section-3

This matches the .NET token type enum: https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsontokentype?view=netcore-3.1#fields